src/Native/Runtime/CommonMacros.h

```diff
@@ -47,17 +49,56 @@ char (*COUNTOF_helper(_CountofType (&_Array)[_SizeOfArray]))[_SizeOfArray];
 #endif
 #endif
 
-inline UIntNative ALIGN_UP(UIntNative val, UIntNative alignment);
-template <typename T>
-inline T* ALIGN_UP(T* val, UIntNative alignment);
+#ifndef __GCENV_BASE_INCLUDED__
 
-inline UIntNative ALIGN_DOWN(UIntNative val, UIntNative alignment);
-template <typename T>
-inline T* ALIGN_DOWN(T* val, UIntNative alignment);
+//
+// This macro returns val rounded up as necessary to be a multiple of alignment; alignment must be a power of 2
+//
+inline size_t ALIGN_UP( size_t val, size_t alignment )
```

Comment:
I would be really nice to keep the implementations out of the .h files, and in the .inl file. Having the implementations in the .h file turns into cyclic dependencies that are very hard to break once they spread around.

Reply:
@jkotas @MichalStrehovsky suggested this change https://github.com/dotnet/corert/commit/15831ea15ef21b1b56bcdecf34a0efa22d932a20 because i've run into this issue:
> so [this commit](https://github.com/dotnet/coreclr/commit/8acf6c9f7e33b1bfe77205cd2e61f2c60e8c00ce#diff-56b3382a15883329e6d34e2411e3e4a0R279) added functions like ALIGN_UP to the GC into a header file  and corert declares them here https://github.com/dotnet/corert/blob/73dcc0508062a266cbada6169fb99566127f3ea5/src/Native/Runtime/CommonMacros.h#L50-L56 and defines them here https://github.com/dotnet/corert/blob/73dcc0508062a266cbada6169fb99566127f3ea5/src/Native/Runtime/CommonMacros.inl#L5-L33 which causes the build to spew about `error C2084: function 'size_t ALIGN_UP(size_t,size_t)' already has a body` so i thought, lets just remove the definition from CommonMacros.inl but that causes linking errors, also removing the declaration causes errors about them missing and including the gcenv.base.h that declares them on the gc side causes different issues...

Reply:
I suggested that because the entire function is in the header in the CoreCLR codebase.

It's possible that having the h/inl split was only problematic because it was a template in CoreRT and C++ templates are a special kind of hell. Maybe making it not a template and still keeping the split would work?

Reply:
@MichalStrehovsky actually, your little trick to check `#ifndef GCENV_INCLUDED` is all that is needed to resolve the original problem 😃